File: MC.py
import itertools
import operator
import math
import numpy
''' ----- Permutation Group System ----- 
attributes: n, elements, order, index, chi'''
class Group(object):
    # construct symmetric group S_n given n
    def __init__(self, n):
        self.n = n
        self.elements = list(itertools.permutations(range(n)))
        self.dof = len(self.elements) # group order
        self.index = {g:i for g, i in zip(self.elements, range(self.dof))}
        self.build_chi() # build chi table
        # after building chi table, self.chi is constructed
    def __repr__(self):
        return repr(self.elements)
    def __iter__(self):
        return iter(self.elements)
    # group multiplication
    def multiply(self, g1, g2):
        return tuple(g1[i] for i in g2)
    # group inversion
    def inverse(self, g):
        return tuple(sorted(range(self.n), key=lambda k: g[k]))
    # permutation character (number of cycles)
    # the range: character =  1,2,...,n
    def character(self, g):
        # setup a set of objects
        objs = set(range(self.n))
        cycles = 0 # initially no cycles
        while objs: # if objects not empty
            obj = objs.pop() # pop out an object
            obj = g[obj] # get its image object under g action
            while obj in objs: # if obj is still in objs
                objs.remove(obj) # remove obj
                obj = g[obj] # and get the next image
            # while loop terminates if a full cycles has been gone through
            cycles += 1 # cycles +1
        return cycles
    # build chi table: chi(g1, g2) = ln Tr g1*g2^(-1)
    def build_chi(self):
        g1s = self.elements
        # prepare the inversions
        g2s = [self.inverse(g) for g in self.elements]
        # prepare the characters adjusted by n, such that chi <= 0
        # the reason to adjust by n is to avoid overflow of partition weight at low temperature
        chis = [self.character(g)-self.n for g in self.elements]
        # construct chi table
        self.chi = [[chis[self.index[self.multiply(g1,g2)]] for g1 in g1s] for g2 in g2s]
''' ----- Lattice System ----- 
attributes: nsite, na, nb, nc, sites, coordmap,
            (Hamiltonian related) irng, jlst, klst, nlst'''
class Lattice(object):
    # initialize an empty lattice
    def __init__(self):
        self.empty()
        self.build_Hamiltonian()
    def empty(self):
        self.nsite = 0 # number of lattice sites
        self.na, self.nb, self.nc = 0,0,0 # sublattice sites partition
        self.sites = [] # lattice sites (container)
        self.coordmap = {} # coordinate map: coordinate tuple => site object
    def __repr__(self):
        return repr(self.sites)
    # for site in Lattice will iterate over all sites
    def __iter__(self):
        return iter(self.sites)
    # Lattice(coord) will return the site specified by the coordinate tuple
    def __getitem__(self, coord):
        return self.coordmap[coord]
    # build_Hamiltonian returns the tuple (irng, jlst, klst, nlst) as numpy array
    # which will be passed to FORTRAN MC extension core in the Model class
    # From these data, Hamiltonian is constructed as (in a row compressed fassion)
    #     slice(i) := irng[i]:irng[i+1]-1
    #     H = - sum_i [chi(g(i),g(j)) for j in jlst[slice(i)]] dot klst[slice(i)]
    #     nlst specifies the length of jlst and klst
    # Here we provide a phantom version for null Hamiltonian
    # Note: the inheritant class must redefine the build_Hamiltonian method!
    def build_Hamiltonian(self):
        self.irng = numpy.array([1]*(self.nsite+1))
        self.jlst = numpy.array([],dtype=numpy.int_)
        self.klst = numpy.array([],dtype=numpy.float_)
        self.nlst = 0
# Inheritant class: HypertreeLattice
class HypertreeLattice(Lattice):
    # Site class specific to HypertreeLattice
    class Site(object):
        # site constructor given (d,x,f)
        def __init__(self, d, x, f):
            self.d = d # depth
            self.f = f # feature space coordinate
            self.x = x # real space coordinate
            self.UV = [] # UV sites
            self.IR = [] # IR sites
            self.label = None
            self.index = None
        def __repr__(self):
            return '<%d,%d,%d>'%(self.d,self.f,self.x)
        def sortkey(self):
            return (self.label,-self.d,self.f,self.x)
    # lattice constructor given UV system size L
    def __init__(self, L=0, Ks=[], branches=2):
        Lattice.empty(self) # setup an empty lattice
        self.L = L # keep system size L
        self.branches = branches # set branches at each RG step
        # UV layer of sites
        layer0 = [self.Site(0,x,0) for x in range(L)]
        self.ftop = {} # initialize feature top register
        # build hypertree
        self.depth = 0
        self.sites = self.hypertree(layer0)
        self.nsite = len(self.sites)
        # on return self.depth has been set to the depth of the tree
        # label sites: A - A sublattice, B - B sublattice, C - boundary
        for site in self.sites:
            if site.d == 0: # the ultimate UV layer is the boundary
                site.label = 'C'
                self.nc += 1
            elif site.d%2 == self.depth%2:
                # A sublattice must contain the ultimate IR layer
                site.label = 'A'
                self.na += 1
            else: # B sublattice contains the rest
                site.label = 'B'
                self.nb += 1
        # sort sites and setup coordinate mapping
        self.sites.sort(key = operator.methodcaller('sortkey'))
        for index, site in zip(range(self.nsite), self.sites):
            site.index = index
            self.coordmap[(site.d,site.f,site.x)] = site
        # build Hamiltonian, after building,
        # self.irng,.jlst,.klst,.nlst will be set
        self.build_Hamiltonian(Ks)
    # recursive build hypertree
    def hypertree(self, this_layer, d=1):
        # if this_layer has exausted
        if len(this_layer) <= 1:
            self.depth = max(self.depth, d-1)
            return this_layer
        # get the top index of features in this depth
        try:
            f = self.ftop[d]
        except: # if failed, set new feature top
            self.ftop[d] = 0
            f = 0
        # prepare next layers
        next_layers = []
        x = 0 # initialized next layer real space
        df = 0 # number of new features
        # partition this_layer by into groups of length = branches
        for group in itertools.zip_longest(*[iter(this_layer)]*self.branches):
            # partition UV sites into UV groups
            grp_UV = [site for site in group if site is not None]
            # construct IR sites of the same size as UV in the group
            grp_IR = [self.Site(d,x,f+f1) for f1 in range(len(grp_UV))]
            #print('%d,%d,%d'%(d,x,f),grp_UV,grp_IR)
            # link UV and IR sites
            for site_UV, site_IR in itertools.product(grp_UV, grp_IR):
                site_UV.IR.append(site_IR)
                site_IR.UV.append(site_UV)
            # add IR sites to the next layers
            next_layers.append(grp_IR)
            x += 1
            df = max(df, len(grp_UV))
        # ftop increase by the number of features developed in this step
        self.ftop[d] += df
        sites = this_layer
        for layer_raw in itertools.zip_longest(*next_layers):
            layer = [site for site in layer_raw if site is not None]
            sites.extend(self.hypertree(layer, d+1))
        return sites
    # build Hamiltonian
    def build_Hamiltonian(self, Ks):
        # energy parameters are passed in from Ks (as list or array)
        H = [] # container for Hamiltonian terms
        # each Hamiltonian term is a tuple (i,j,K)
        for site in self.sites:
            i = site.index # take site index
            IRsites = site.IR # get its IR sites
            for IRsite, K in zip(IRsites, Ks):
                j = IRsite.index # take IR site index
                H.append((i,j,K))
                H.append((j,i,K))
        if H:
            ilst, jlst, klst = tuple(zip(*sorted(H)))
        else:
            ilst, jlst, klst = [], [], []
        nlst = len(ilst)
        # now ilst, jlst, klst separately hold (i,j,K) for all terms
        # ilst has been sorted in assending order, convert to irng
        imax = 0
        irng = [0]
        p = -1
        for p, i in zip(range(nlst),ilst):
            while i > imax:
                irng.append(p)
                imax += 1
        irng.extend([p+1]*(self.nsite-imax))
        # return tuple (irng, jlst, klst)
        # Note: +1 is needed to convert irng and jlst to FORTRAN index
        #       missing +1 will lead to bus error in the runtime
        self.irng = numpy.array(irng)+1
        self.jlst = numpy.array(jlst)+1
        self.klst = numpy.array(klst)
        self.nlst = nlst
''' ----- Model System -----
attributes:
    system: dof, nsite, na, nb, nc, nlst, chi, irng, jlst, klst
    state: energy, config, hist
    data: nspin, monitor, energy1, energy2, magnet1, magnet2, spins
methods: run(), measure()
Generic statistical mechanics model:
* onsite dof: g in range(model.dof)
* Hamiltonian H = - sum_{ij} K_{ij} chi(g(i), g(j))'''
import MC # FORTRAN extension: Monte Carlo kernel
class Model(object):
    # specify system parameters
    system_parameters = ('dof','nsite','na','nb','nc','nlst','chi','irng','jlst','klst')
    # Model constructor given system parameter
    # system must be a dict containing keys specified in Model._system_parameters
    def __init__(self, system, state={}, data={}):
        # passing system parameters to MC.core
        self.system = system
        # kernel initialization
        MC.core.init() # private workspace allocated
        # initialize state parameters
        if state == {}:
            self.state = {'config':'FM','energy':'unknown','hist':'unknown'}
        else:
            self.state = state
        self.data = data
    def __getattr__(self, attrname):
        if attrname == 'config':
            if MC.core.config is None:
                return None
            else:
                return MC.core.config-1 # shift back to python index convesion
        elif attrname == 'energy':
            if numpy.isnan(MC.core.energy):
                MC.core.get_energy()
            return numpy.asscalar(MC.core.energy)
        elif attrname == 'hist':
            if MC.core.hist[0] < 0:
                MC.core.get_hist()
            return MC.core.hist
        elif attrname in Model.system_parameters:
            return getattr(MC.core, attrname)
        elif attrname == 'state':
            return {key:getattr(self,key) for key in ('config','energy','hist')}
        elif attrname == 'system':
            return {key:getattr(self,key) for key in Model.system_parameters}
        else:
            raise AttributeError("'%s' object has no attribute '%s'"%(self.__class__.__name__, attrname))
    def __setattr__(self, attrname, attrval):
        if attrname == 'config':
            if isinstance(attrval, numpy.ndarray):
                MC.core.config = attrval+1 # shift to FORTRAN index convension
            elif isinstance(attrval, (list, tuple)):
                MC.core.config = numpy.array(attrval)+1 # shift to FORTRAN index convension
            elif attrval in {'FM','uniform'}:
                MC.core.config = numpy.full(self.nsite,1,dtype=numpy.int_)
            elif attrval in {'PM','random'}:
                MC.core.config = numpy.random.randint(self.dof, size=self.nsite)+1
            else:
                raise ValueError('Illigal value %s for config.'%repr(attrval))
        elif attrname == 'energy':
            if attrval in {'unknown', 'nan'}:
                MC.core.energy = numpy.nan
            else:
                MC.core.energy = attrval
        elif attrname == 'hist':
            if attrval in {'unknown'}:
                MC.core.hist = numpy.empty(self.dof)
                MC.core.hist[0] = -1
            else:
                MC.core.hist = attrval
        elif attrname in Model.system_parameters:
            setattr(MC.core, attrname, attrval)
        elif attrname in {'state', 'system'}:
            for key in attrval:
                setattr(self, key, attrval[key])
        else:
            self.__dict__[attrname] = attrval
    # run MC for steps, under mode 0 or 1
    def run(self, steps=1, mode=0):
        if mode==0: # MC without update physical observibles
            MC.core.run(steps, 0)
        elif mode==1: # MC with physical observibles updated
            MC.core.run(steps, 1)
        else: # unrecognised mode
            raise ValueError('The mode %s of Model.run should be 0 or 1.'%repr(mode))
        return self
    # take measurement for steps, monitoring specified spins
    def measure(self, steps=1, monitor=None):
        if monitor is None:
            MC.physics.nspin = 0
            MC.physics.monitor = numpy.array([],dtype=numpy.int_)
        else:
            if all(0<= i < self.lattice.nsite for i in monitor):
                MC.physics.nspin = len(monitor)
                MC.physics.monitor = numpy.array(monitor)+1
            else:
                raise ValueError('The monitor %s of Model.measure out of lattice range.'%repr(monitor))
        MC.physics.measure(steps)
        self.data = {
            'energy1': numpy.asscalar(MC.physics.energy1), # energy 1st moment
            'energy2': numpy.asscalar(MC.physics.energy2), # energy 2nd moment
            'magnet1': MC.physics.magnet1, # bulk magnetization 1st moment
            'magnet2': MC.physics.magnet2, # bulk magnetization 2nd moment 
            'spins':MC.physics.spins,'steps': steps # monitored spin states
        }
        return self.data
# Inheritant class: LatticeModel - construct model from Lattice and Group
# LatticeModel has two more attributes: lattice and group
class LatticeModel(Model):
    def __init__(self, lattice=Lattice(), group=Group(1)):
        # save lattice and group as attribute
        self.lattice = lattice
        self.group = group
        # construct system parameters
        para = {key:val for key,val in itertools.chain(lattice.__dict__.items(),group.__dict__.items())}
        system = {name:para[name] for name in Model.system_parameters}
        # call Model method to initialize
        Model.__init__(self, system)

File: MC.f90
! TO COMPILE -----------
! 1. create signature file by:
!    f2py MC.f90 -m MC -h MC.pyf --overwrite-signature
! 2. remove private variables and procedures in MC.pyf
! 3. build extension by:
!    f2py -c MC.pyf MC.f90 --opt='-O3' --build-dir build
! +-----------------------------------+
! |   Fortran core for Monte Carlo    |
! +-----------------------------------+
MODULE CORE
    USE IEEE_ARITHMETIC
    IMPLICIT NONE
    ! data pool linked to python
    ! system variables
    INTEGER :: DOF    ! onsite degrees of freedom
    INTEGER :: NSITE  ! number of sites
    INTEGER :: NA, NB, NC ! number of sublattice sites
    INTEGER :: NLST   ! size of JLST and KLST
    INTEGER, ALLOCATABLE :: CHI(:,:)  ! chi table
    INTEGER, ALLOCATABLE :: IRNG(:)   ! adjacency range list
    INTEGER, ALLOCATABLE :: JLST(:)   ! adjacent site index list
    REAL,    ALLOCATABLE :: KLST(:)   ! action coefficient list
    ! state variables
    REAL :: ACTION ! action
    REAL :: BETA   ! inverse temperature
    INTEGER, ALLOCATABLE :: CONFIG(:) ! configuration    
    INTEGER, ALLOCATABLE :: HIST(:)   ! histogram of spin (bin count)
    ! private workspace
    INTEGER :: NBLK ! number of bulk sites
    REAL, ALLOCATABLE :: BIAS(:,:)   ! local bias field
    REAL, ALLOCATABLE :: WEIGHT(:,:) ! local weights
    REAL, ALLOCATABLE :: RND(:)      ! random numbers
    REAL, ALLOCATABLE :: KLST1(:)    ! an initial copy KLST for tempering
CONTAINS
    ! ------ Initialization ------
    ! initialization
    SUBROUTINE INIT()
        ! allocate workspace
        ! bulk sites = A + B sites
        NBLK = NA + NB
        IF (ALLOCATED(BIAS)) THEN
            IF (ANY(SHAPE(BIAS) /= [DOF, NBLK])) THEN
                DEALLOCATE(BIAS)
                ALLOCATE(BIAS(DOF, NBLK))
            END IF
        ELSE
            ALLOCATE(BIAS(DOF, NBLK))
        END IF
        IF (ALLOCATED(WEIGHT)) THEN
            IF (ANY(SHAPE(WEIGHT) /= [DOF, NBLK])) THEN
                DEALLOCATE(WEIGHT)
                ALLOCATE(WEIGHT(DOF, NBLK))
            END IF
        ELSE
            ALLOCATE(WEIGHT(DOF, NBLK))
        END IF
        IF (ALLOCATED(RND)) THEN
            IF (ANY(SHAPE(RND) /= [NBLK])) THEN
                DEALLOCATE(RND)
                ALLOCATE(RND(NBLK))
            END IF
        ELSE
            ALLOCATE(RND(NBLK))
        END IF
        ! random seed
        CALL INIT_RAND_SEED()
        ! make an initial copy of KLST
        ! it is better to update KLST from its initial copy
        ! to avoid error accumulation after several updates
        IF (NLST > 0) THEN
            KLST1 = KLST
        END IF
    END SUBROUTINE INIT
    ! automatic random seed
    SUBROUTINE INIT_RAND_SEED()
        INTEGER :: I, N, CLOCK
        INTEGER, DIMENSION(:), ALLOCATABLE :: SD
        CALL RANDOM_SEED(SIZE = N)
        ALLOCATE(SD(N))
        CALL SYSTEM_CLOCK(COUNT = CLOCK)
        SD = CLOCK + 37*[(I-1,I=1,N)]
        CALL RANDOM_SEED(PUT = SD)
        DEALLOCATE(SD)
    END SUBROUTINE INIT_RAND_SEED
    ! manually set random seed
    SUBROUTINE SEED(S)
        INTEGER, INTENT(IN) :: S
        INTEGER :: I, N
        INTEGER, DIMENSION(:), ALLOCATABLE :: SD
        CALL RANDOM_SEED(SIZE = N)
        ALLOCATE(SD(N))
        SD = S + 37*[(I-1,I=1,N)]
        CALL RANDOM_SEED(PUT = SD)
        DEALLOCATE(SD)
    END SUBROUTINE SEED
    ! ------ Sampling ------
    ! k steps of MC update (each update runs over all sites)
    SUBROUTINE RUN(STEPS)
        INTEGER, INTENT(IN) :: STEPS
        INTEGER :: STEP
        ! run a simple sampling without keeping track of states
        DO STEP = 1, STEPS
            CALL RANDOM_NUMBER(RND) ! prepare RND
            CALL SAMPLE0(1, NA)
            CALL SAMPLE0(NA+1, NA+NB)
        END DO
        ! recalculate the action and hist after running
        CALL GET_ACTION()
        CALL GET_HIST()
    END SUBROUTINE RUN
    ! block Gibbs sampling (mode 0: no update physical variables)
    SUBROUTINE SAMPLE0(LB, UB)
        INTEGER, INTENT(IN) :: LB, UB
        INTEGER :: I
        ! set BIAS, WEIGHT, and adjust RND for random choice
        CALL SET_BLOCK(LB, UB)
        ! make random choice by binary search
        FORALL (I = LB:UB)
            CONFIG(I) = CHOOSE(WEIGHT(:, I),RND(I))
        END FORALL
    END SUBROUTINE SAMPLE0
    ! block Gibbs sampling (mode 1: update physical variables)
    SUBROUTINE SAMPLE1(LB, UB)
        INTEGER, INTENT(IN) :: LB, UB
        INTEGER :: I, G0, G1
        ! set BIAS, WEIGHT, and adjust RND for random choice
        CALL SET_BLOCK(LB, UB)
        ! make random choice by binary search
        DO I = LB, UB
            G0 = CONFIG(I)
            G1 = CHOOSE(WEIGHT(:, I),RND(I))
            CONFIG(I) = G1
            ACTION = ACTION + BIAS(G0, I) - BIAS(G1, I)
            HIST(G0) = HIST(G0) - 1
            HIST(G1) = HIST(G1) + 1
        END DO
    END SUBROUTINE SAMPLE1
    ! calculate accumulated weights
    SUBROUTINE SET_BLOCK(LB, UB)
        INTEGER, INTENT(IN) :: LB, UB
        INTEGER :: I
        ! set BIAS, WEIGHT block, and renormalize RND
        FORALL (I = LB:UB)
            BIAS(:,I) = SET_BIAS(I) ! faster than MATMUL
            WEIGHT(:,I) = SET_WEIGHT(BIAS(:,I)) ! combined
            RND(I) = RND(I) * WEIGHT(DOF, I)            
        END FORALL
    END SUBROUTINE SET_BLOCK
    ! calculate bias field (as a vector of DOF components)
    ! given the site index I
    PURE FUNCTION SET_BIAS(I) RESULT (B)
        INTEGER, INTENT(IN) :: I
        REAL :: B(DOF)
        INTEGER :: J
        B = 0.
        DO J = IRNG(I), IRNG(I+1)-1
            B = B + CHI(:,CONFIG(JLST(J)))*KLST(J)
        END DO
        ! this is several times faster then the MATMUL implementation
        ! because on matrix is constructed before multiplication
    END FUNCTION SET_BIAS
    ! calculate accumulated weight
    ! given the site index I
    PURE FUNCTION SET_WEIGHT(B) RESULT (W)
        REAL, INTENT(IN) :: B(DOF)
        REAL :: W(DOF)
        INTEGER :: K
        W(1) = EXP(B(1))
        ! directly accumulate CDF, not constructing PDF first
        DO K = 2, DOF
            W(K) = W(K-1) + EXP(B(K))
        END DO
    END FUNCTION SET_WEIGHT
    ! random choice
    ! given CDF array W, and random number X
    ! return a choice in [1:DOF]
    PURE FUNCTION CHOOSE(W, X) RESULT (R)
        REAL, INTENT(IN) :: W(DOF) ! weight array
        REAL, INTENT(IN) :: X ! a random number
        INTEGER :: L, R, M
        ! binary search
        L = 0
        R = DOF
        ! if R > L+1 do the search
        DO WHILE (R - L > 1)
            M = (L + R)/2 ! set mid point
            IF (W(M) >= X) THEN
                R = M
            ELSE
                L = M
            END IF
        END DO
        ! now X is determined in the bin (L,R=L+1)
        ! L will be the result of the choice
    END FUNCTION CHOOSE
    ! ------ State Modification ------
    ! set beta and update action coefficients
    SUBROUTINE SET_BETA(NEW_BETA)
        REAL, INTENT(IN) :: NEW_BETA 
        BETA = NEW_BETA ! set new beta
        ! update action coefficients
        IF (NLST > 0) THEN
            KLST = KLST1 * BETA ! from the initial copy
        END IF
        ! action is also changed by resetting temperature
        CALL GET_ACTION() ! recalculate action
    END SUBROUTINE SET_BETA
    ! calculate action
    SUBROUTINE GET_ACTION()
        INTEGER :: I
        ACTION = 0.
        DO I = 1, NSITE ! for all sites
            ACTION = ACTION - DOT_PRODUCT(CHI(CONFIG(I), &
                     CONFIG(JLST(IRNG(I):IRNG(I+1)-1))), &
                     KLST(IRNG(I):IRNG(I+1)-1))
        END DO
        ! each bond is doubled counted, so 1/2
        ACTION = ACTION/2
    END SUBROUTINE GET_ACTION 
    ! collect total spin (histogram)
    SUBROUTINE GET_HIST()
        INTEGER :: I, G
        HIST = 0
        DO I = 1, NBLK ! for bulk sites only
            G = CONFIG(I)
            HIST(G) = HIST(G) + 1
        END DO
    END SUBROUTINE GET_HIST
    ! ------ I/O System ------
    ! core dump
    SUBROUTINE DUMP()        
        PRINT *, "MC.core dump to MC.core.dat"
        OPEN (UNIT=99, FILE="MC.core.dat", STATUS="REPLACE", ACCESS="STREAM")
        WRITE(99) DOF, NSITE, NA, NB, NC, NLST, NBLK
        WRITE(99) CHI, IRNG, JLST, KLST
        WRITE(99) ACTION, BETA, CONFIG, HIST
        WRITE(99) BIAS, WEIGHT, RND, KLST1
        CLOSE(99)
    END SUBROUTINE DUMP
    ! core load
    SUBROUTINE LOAD() 
        PRINT *, "MC.core load from MC.core.dat"
        OPEN (UNIT=99, FILE="MC.core.dat", STATUS="UNKNOWN", ACCESS="STREAM")
        READ(99) DOF, NSITE, NA, NB, NC, NLST, NBLK
        IF (ALLOCATED(CHI)) DEALLOCATE(CHI)
        IF (ALLOCATED(IRNG)) DEALLOCATE(IRNG)
        IF (ALLOCATED(JLST)) DEALLOCATE(JLST)
        IF (ALLOCATED(KLST)) DEALLOCATE(KLST)
        IF (ALLOCATED(CONFIG)) DEALLOCATE(CONFIG)
        IF (ALLOCATED(HIST)) DEALLOCATE(HIST)
        IF (ALLOCATED(BIAS)) DEALLOCATE(BIAS)
        IF (ALLOCATED(WEIGHT)) DEALLOCATE(WEIGHT)
        IF (ALLOCATED(RND)) DEALLOCATE(RND)
        IF (ALLOCATED(KLST1)) DEALLOCATE(KLST1)
        ALLOCATE(CHI(DOF,DOF), IRNG(NSITE+1), JLST(NLST), KLST(NLST), &
                 CONFIG(NSITE), HIST(DOF), &
                 BIAS(DOF,NBLK), WEIGHT(DOF,NBLK), RND(NBLK), &
                 KLST1(NLST))
        READ(99) CHI, IRNG, JLST, KLST
        READ(99) ACTION, BETA, CONFIG, HIST
        READ(99) BIAS, WEIGHT, RND, KLST1
        CLOSE(99)
    END SUBROUTINE LOAD
END MODULE CORE
! Measurement module
MODULE DATA
    USE CORE
    IMPLICIT NONE
    ! data variables
    INTEGER :: NSPIN ! number of spins to monitor
    INTEGER, ALLOCATABLE :: MONITOR(:) ! sites for monitoring
    REAL :: ENERGY1, ENERGY2
    REAL, ALLOCATABLE :: MAGNET1(:), MAGNET2(:,:) ! (DOF, DOF)
    REAL, ALLOCATABLE :: SPINS(:,:) ! (DOF, NSPIN)
    ! time series
    INTEGER :: NTS ! length of the time series
    REAL, ALLOCATABLE :: ETS(:)    ! energy density series
    REAL, ALLOCATABLE :: MTS(:,:) ! magnetization density series
CONTAINS
    ! measure energy and spin data over steps
    SUBROUTINE MEASURE(STEPS)
        INTEGER, INTENT(IN) :: STEPS
        INTEGER :: STEP, I, J
        REAL :: ENERGY, MAGNET(DOF)
        
        ! launch measurement environment
        ! magnetization 1st moment (vector)
        IF (ALLOCATED(MAGNET1)) THEN
            IF (ANY(SHAPE(MAGNET1) /= [DOF])) THEN
                DEALLOCATE(MAGNET1)
                ALLOCATE(MAGNET1(DOF))
            END IF
        ELSE
            ALLOCATE(MAGNET1(DOF))
        END IF
        ! magnetization 2nd moment (matrix)
        IF (ALLOCATED(MAGNET2)) THEN
            IF (ANY(SHAPE(MAGNET2) /= [DOF, DOF])) THEN
                DEALLOCATE(MAGNET2)
                ALLOCATE(MAGNET2(DOF, DOF))
            END IF
        ELSE
            ALLOCATE(MAGNET2(DOF, DOF))
        END IF
        ! spin 1st moment (array of vectors)
        ! all higher order moments are the same as 1st moment
        IF (ALLOCATED(SPINS)) THEN
            IF (ANY(SHAPE(SPINS) /= [DOF, NSPIN])) THEN
                DEALLOCATE(SPINS)
                ALLOCATE(SPINS(DOF, NSPIN))
            END IF
        ELSE
            ALLOCATE(SPINS(DOF, NSPIN))
        END IF
        ! get energy and hist ready
        CALL GET_ACTION()
        CALL GET_HIST()
        ! clear data pool
        ENERGY1 = 0.
        ENERGY2 = 0.
        MAGNET1 = 0.
        MAGNET2 = 0.
        SPINS = 0.
        ! run MC with updates, and collect data
        DO STEP = 1, STEPS
            CALL RANDOM_NUMBER(RND) ! prepare RND
            CALL SAMPLE1(1, NA)
            CALL SAMPLE1(NA+1, NA+NB)
            ! energy measurement
            ENERGY = ACTION/BETA/NSITE
            ENERGY1 = ENERGY1 + ENERGY
            ENERGY2 = ENERGY2 + ENERGY**2
            ! magnetization measurement
            MAGNET = REAL(HIST)/NBLK
            MAGNET1 = MAGNET1 + MAGNET
            FORALL (I = 1:DOF, J = 1:DOF)
                MAGNET2(I, J) = MAGNET2(I, J) + MAGNET(I)*MAGNET(J)
            END FORALL
            ! spin measurement
            FORALL (I = 1:NSPIN)
                SPINS(CONFIG(MONITOR(I)), I) = SPINS(CONFIG(MONITOR(I)), I) + 1.
            END FORALL
        END DO
        ! normalized by STEPS
        ENERGY1 = ENERGY1/STEPS
        ENERGY2 = ENERGY2/STEPS
        MAGNET1 = MAGNET1/STEPS
        MAGNET2 = MAGNET2/STEPS
        SPINS = SPINS/STEPS
    END SUBROUTINE MEASURE
    ! collect time series
    SUBROUTINE COLLECT(STEPS)
        INTEGER, INTENT(IN) :: STEPS
        INTEGER :: T
        NTS = STEPS
        ! reallocation for ETS, MTS
        IF (ALLOCATED(ETS).OR.ALLOCATED(MTS)) THEN
            IF (ANY(SHAPE(ETS) /= [NTS]) .OR. &
                ANY(SHAPE(MTS) /= [DOF,NTS])) THEN
                IF (ALLOCATED(ETS)) DEALLOCATE(ETS)
                IF (ALLOCATED(MTS)) DEALLOCATE(MTS)
                ALLOCATE(ETS(NTS), MTS(DOF, NTS))
            END IF
        ELSE
            ALLOCATE(ETS(NTS), MTS(DOF, NTS))
        END IF
        ! get energy and hist ready
        CALL GET_ACTION()
        CALL GET_HIST()
        ! start collect time series
        DO T = 1, NTS
            CALL RANDOM_NUMBER(RND) ! prepare RND
            CALL SAMPLE1(1, NA)
            CALL SAMPLE1(NA+1, NA+NB)
            ETS(T) = ACTION/BETA/NSITE
            MTS(:,T) = REAL(HIST)/NBLK
        END DO
    END SUBROUTINE COLLECT
END MODULE DATA

! ========== for debug use ==========
! SUBROUTINE TEST_CHOOSE()
!     USE CORE
!     INTEGER :: L
!     DOF = 6
!     L = CHOOSE([2.,3.,7.,15.,20.,23.],22.99)
!     PRINT *, L
! END SUBROUTINE TEST_CHOOSE
! SUBROUTINE TEST_RUN()
!     USE CORE
!     !CALL RUN(1,1)
!     PRINT *, CONFIG
!     CALL GET_ACTION()
!     PRINT *, ACTION
! END SUBROUTINE TEST_RUN
! SUBROUTINE TEST_MEASURE()
!     USE DATA
!     CALL INIT_RAND_SEED()
!     CALL MEASURE(100)
!     PRINT *, NSPIN
!     PRINT *, ENERGY1, ENERGY2
!     PRINT *, MAGNET1
!     PRINT *, CONFIG
! END SUBROUTINE TEST_MEASURE
! SUBROUTINE TEST_SETBETA()
!     USE CORE
!     PRINT *, BETA, ACTION
!     PRINT *, KLST(1:3),'...'
!     PRINT *, KLST1(1:3),'...'
!     CALL SET_BETA(1.)
!     PRINT *, BETA, ACTION
!     PRINT *, KLST(1:3),'...'
!     PRINT *, KLST1(1:3),'...'
! END SUBROUTINE TEST_SETBETA
! 
! PROGRAM MAIN
!     USE CORE
!     
!     CALL LOAD()
!     CALL TEST_SETBETA()
! !     CALL TEST_RUN()
! !     CALL TEST_MEASURE()
! !     CALL TEST_CHOOSE()
! END PROGRAM MAIN